beetsplug/web/__init__.py

```diff
@@ -327,7 +327,7 @@ def item_file(item_id):
 
     try:
         # Imitate http.server behaviour
-        base_filename.encode("latin-1", "strict")
+        base_filename.decode().encode("latin-1", "strict")
```

Comment:
Could you explain this change? I think `base_filename` should be `str` here, but mypy might not be able to infer this. In any case, I don't understand what this code does at all, it might need a proper cleanup.

Reply:
This has to do with the change I made in `syspath` 
![image](https://github.com/beetbox/beets/assets/16212750/03f55a15-dae8-430c-b57e-b55a8903fe38)

You can see that it now **always** returns the same type that was fed into it. 

```python
    return str_path if isinstance(path, str) else str_path.encode()
```

It previously returned undeterministic `BytesOrStr`: for non-windows systems it would return the same type that was fed in, but on Windows it would always return `str`. This caused a fair bit of typing issues down the line.

Now the code in `web/__init__.py` is a bit of a mess

```python
    # On Windows under Python 2, Flask wants a Unicode path. On Python 3, it
    # *always* wants a Unicode path.
    if os.name == "nt":
        item_path = util.syspath(item.path)
    else:
        item_path = util.py3_path(item.path)

    base_filename = os.path.basename(item_path)
    # FIXME: Arguably, this should just use `displayable_path`: The latter
    # tries `_fsencoding()` first, but then falls back to `utf-8`, too.
    if isinstance(base_filename, bytes):
        try:
            unicode_base_filename = base_filename.decode("utf-8")
        except UnicodeError:
            unicode_base_filename = util.displayable_path(base_filename)
    else:
        unicode_base_filename = base_filename

    try:
        # Imitate http.server behaviour
        base_filename.decode().encode("latin-1", "strict")
```

You can see it calls `syspath` on Windows, which now returns `bytes` instead of `str`. It seemingly expects `base_filename` to have type `bytes` and assigns `unicode_base_filename`, but then ends up using `base_filename` in the last line here, assuming that it's a `str` but it is still `bytes`.

Reply:
Given this comment

```python
    # On Windows under Python 2, Flask wants a Unicode path. On Python 3, it
    # *always* wants a Unicode path.
    if os.name == "nt":
        item_path = util.syspath(item.path)
    else:
        item_path = util.py3_path(item.path)
```

I think it should be safe to simply call `py3_path` in all cases now.

Reply:
I agree that `syspath` is a tricky regarding typing — but this is in fact a change which I think has the potential to cause a big fallout if we get it wrong, and at this point I'm not fully convinced that it is correct.

Up to now, `syspath` is intended to return `str` on Windows, because all Windows filesystems and APIs use some Unicode encoding and the previous default `mbcs` encoding (which I think would have been used when passing `bytes` to OS APIs) would in fact have been incorrect, given that beets internally uses UTF-8 as stated in the code comment.

Due to https://peps.python.org/pep-0529/ your change might in fact be correct now (note that there's a legacy mode that uses MBCS, even though I'd say users who explicitly enable that are on their own).

I'd prefer not to change the behaviour for now (at least not in this PR, let's keep the code changes to a minimum when the primary goal is introducing typings). I'm not confident that our testsuite exercises non-ASCII filenames sufficiently to catch all encoding edge cases.

For now, another solution might be to define types conditionally (mypy understands the `sys.platform` switch IIRC):

```python
if sys.platform == "win32":
    SysPath: TypeAlias = str
else:
    SysPath: TypeAlias = bytes
def syspath(path: bytes, prefix: bool = True) -> SysPath:
```
(we should probably also run mypy on Windows then).


---

Regarding the `web` plugin code: If I'm not completely mistaken, `base_filename` is guaranteed to be `str` when `decode` is invoked (irrespective of the `syspath` change, in fact), but only `bytes` can be `decode()`ed.

Actually, it looks like the whole section
```python
    base_filename = os.path.basename(item_path)
    # FIXME: Arguably, this should just use `displayable_path`: The latter
    # tries `_fsencoding()` first, but then falls back to `utf-8`, too.
    if isinstance(base_filename, bytes):
        try:
            unicode_base_filename = base_filename.decode("utf-8")
        except UnicodeError:
            unicode_base_filename = util.displayable_path(base_filename)
    else:
        unicode_base_filename = base_filename
```
is a no-op nowadays (always hitting the `else` arm. (It's not hit in tests at least: https://app.codecov.io/gh/beetbox/beets/blob/master/beetsplug%2Fweb%2F__init__.py but note that there, also the `UnicodeError` never happens, i.e. this is a place where we don't seem to actually test non-ASCII (actually, non-latin1) filenames.)

Reply:
Actually, the `decode` is where the tests fail.

Reply:
> For now, another solution might be to define types conditionally (mypy understands the `sys.platform` switch IIRC):
> 
> ```python
> if sys.platform == "win32":
>     SysPath: TypeAlias = str
> else:
>     SysPath: TypeAlias = bytes
> def syspath(path: bytes, prefix: bool = True) -> SysPath:
> ```

This is a nice solution - though the catch here is that for non-win32 systems the function actually returns `AnyStr`, and I'm not sure how to type it to avoid this issue

![image](https://github.com/beetbox/beets/assets/16212750/ddf8401a-1d8f-4fcf-9a08-634e2d000abf)

Regarding the failing tests, oops, I was meant to use `os.fsdecode` there

Reply:
> Actually, it looks like the whole section
> 
> ```python
>     base_filename = os.path.basename(item_path)
>     # FIXME: Arguably, this should just use `displayable_path`: The latter
>     # tries `_fsencoding()` first, but then falls back to `utf-8`, too.
>     if isinstance(base_filename, bytes):
>         try:
>             unicode_base_filename = base_filename.decode("utf-8")
>         except UnicodeError:
>             unicode_base_filename = util.displayable_path(base_filename)
>     else:
>         unicode_base_filename = base_filename
> ```
> 
> is a no-op nowadays (always hitting the `else` arm. (It's not hit in tests at least: https://app.codecov.io/gh/beetbox/beets/blob/master/beetsplug%2Fweb%2F__init__.py

Note that the coverage is produced by running the tests in the `ubuntu` machine, so it doesn't show the logical branches that Windows hits 🙁

Reply:
Regarding `syspath`, I reckon we could use the good ol' `cast`?

```python
    return cast(AnyStr, str_path)
```

Reply:
`syspath` implementation is now back to where it was before